palace/utils/geodata.cpp

```diff
@@ -993,95 +992,191 @@ BoundingBox BoundingBoxFromPointCloud(MPI_Comm comm,
-    if (!ball.planar)
-    {
-      // The points are not functionally coplanar, zero out the normal.
-      MFEM_VERIFY(std::abs(PerpendicularDistance({delta}, perp) - ball.radius) <=
-                      rel_tol * ball.radius,
-                  "Furthest point perpendicular must be on the exterior of the sphere!");
-      Vector3dMap(ball.planar_normal.data()) *= 0;
-    }
+  return D;
+}
+
+// Calculates a bounding ball from a point cloud using Welzl's algorithm, result is
+// broadcast across all processes. We don't operate on the convex hull, since the number of
+// points should be small enough that operating on the full set should be OK. If only three
+// points are provided, the bounding circle is computed (likewise for if the points are
+// coplanar).
+BoundingBox BoundingBallFromPointCloud(MPI_Comm comm,
+                                       const std::vector<Eigen::Vector3d> &vertices,
+                                       int dominant_rank)
+{
+  BoundingBox ball;
+  if (dominant_rank == Mpi::Rank(comm))
+  {
+    // Randomly permute the point set.
+    MFEM_VERIFY(vertices.size() >= 3,
+                "A bounding ball requires a minimum of three vertices for this algorithm!");
+    std::vector<std::size_t> indices(vertices.size() + 4);
+    for (std::size_t i = 0; i < vertices.size(); i++)
+    {
+      indices[i] = i;
+    }
+    {
+      std::random_device rd;
+      std::mt19937 g(rd());
+      std::shuffle(indices.begin(), indices.end() - 4, g);
+    }
+
+    // Acceleration from https://informatica.vu.lt/journal/INFORMATICA/article/1251. Allow
+    // for duplicate points and just add the 4 points to the end of the indicies list to be
+    // considered first. The two points are not necessarily the maximizer of the distance
+    // between all pairs, but they should be a good estimate.
+    {
+      auto p_1 = std::min_element(vertices.begin(), vertices.end(), EigenLE);
+      auto p_2 = std::max_element(vertices.begin(), vertices.end(),
+                                  [p_1](const Eigen::Vector3d &x, const Eigen::Vector3d &y)
+                                  { return (x - *p_1).norm() < (y - *p_1).norm(); });
+      p_1 = std::max_element(vertices.begin(), vertices.end(),
+                             [p_2](const Eigen::Vector3d &x, const Eigen::Vector3d &y)
+                             { return (x - *p_2).norm() < (y - *p_2).norm(); });
+      auto p_12 = 0.5 * (*p_1 + *p_2);
+      auto p_3 =
+          std::max_element(vertices.begin(), vertices.end(),
+                           [&p_12](const Eigen::Vector3d &x, const Eigen::Vector3d &y)
+                           { return (x - p_12).norm() < (y - p_12).norm(); });
+      auto p_4 = std::max_element(vertices.begin(), vertices.end(),
+                                  [p_3](const Eigen::Vector3d &x, const Eigen::Vector3d &y)
+                                  { return (x - *p_3).norm() < (y - *p_3).norm(); });
```

Comment:
I'll just revert to the projected distance, since it's already implemented. Not difficult to do and should be an improvement even if it doesn't matter much if at all for performance.

Reply:
Done in [32ebd0c](https://github.com/awslabs/palace/pull/246/commits/32ebd0c207e95e6a463007517689f75af0240dd1).